enarx-keep-sgx-shim/src/handler.rs

```diff
@@ -254,31 +253,29 @@ impl<'a> Handler<'a> {
     pub fn read(&mut self) -> u64 {
         self.trace("read", 3);
 
-        let fd = self.aex.gpr.rdi.raw();
-        let buf = self.aex.gpr.rsi.raw() as *mut u8;
-        let size = self.aex.gpr.rdx.raw();
+        if usize::from(self.aex.gpr.rdx) > self.block.buf.len() {
+            return -libc::EMSGSIZE as u64;
+        }
 
-        // Allocate some unencrypted memory.
-        let map = match self.ualloc(size) {
-            Err(errno) => return errno as u64,
-            Ok(map) => map,
+        // Do the syscall; replace encrypted memory with unencrypted memory.
+        let res = unsafe {
+            self.proxy(request!(libc::SYS_read => self.aex.gpr.rdi, &self.block.buf, usize::from(self.aex.gpr.rdx)))
```

Comment:
```suggestion
            self.proxy(request!(libc::SYS_read => self.aex.gpr.rdi, &self.block.buf, self.aex.gpr.rdx))
```

This should work. If not, we probably have a bug in `Register` conversions. `self.aex.gpr.rdx` should be `Register<u64>`, which should infallibly convert to `Register<usize>` on 64-bit systems.